### PR TITLE
Feature/include profiles

### PR DESCRIPTION
**Pull request by lasote**
#1145

### CODE REVIEW

File: conans/test/model/profile_test.py [hunk 1]
```diff
@@ -11,54 +12,12 @@
...
-                           ConanFileReference.loads("autotools/1.0.3@lasote/stable")])
-        self.assertEquals(new_profile.build_requires["*"],
-                          [ConanFileReference.loads("android_toolchain/1.2.8@lasote/testing")])
 
     def profile_settings_update_test(self):
         prof = '''[settings]
 os=Windows
 '''
-        new_profile = Profile.loads(prof)
+        new_profile, _ = _load_profile(prof, None, None)
```
Comment (memsharded) [inline conans/test/model/profile_test.py]: testing private methods?

File: conans/test/functional/profile_loader_test.py [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+
+    def setUp(self):
+        self.client = TestClient()
+
+    def profile_loads_test(self):
+        prof = '''[env]
+    CXX_FLAGS="-DAAA=0"
+    [settings]
+    '''
+        new_profile, _ = _load_profile(prof, None, None)
```
Comment (memsharded) [inline conans/test/functional/profile_loader_test.py]: Testing privates

File: conans/client/profile_loader.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        for name, value in local_vars.items():
+            tmp_block_text = tmp_block_text.replace("$%s" % name, value)
+
+        self.text = self._defs_block_text + "\n" + tmp_block_text
+
+    @property
+    def includes(self):
+        ret = []
+        for line in self._defs_block_text.splitlines():
+            if not line or line.strip().startswith("#"):
```
Comment (memsharded) [inline conans/client/profile_loader.py]: These have already been considered in ``_defs_block_text``

File: conans/client/profile_loader.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        """Returns the text for #includes and vars declaration"""
+        ret_lines = []
+        for line in self.text.splitlines():
+            if not line or line.strip().startswith("#"):
+                continue
+            elif line.strip().startswith("["):
+                return "\n".join(ret_lines)
+            else:
+                ret_lines.append(line)
+        return "\n".join(ret_lines)
```
Comment (memsharded) [inline conans/client/profile_loader.py]: This conversion to text again is confusing, then parsing them again line by line for includes and vars.

File: conans/client/profile_loader.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+    @property
+    def config_block_text(self):
+        """Returns the text for pure profile"""
+        if not "[" in self.text:
+            return ""
+        return "[" + self.text.split("[", 1)[1]
```
Comment (memsharded) [inline conans/client/profile_loader.py]: this reads very difficult to understand what is going on

**Comment by memsharded** [comment]
I'd say that the recursive parsing and substitution is difficult to understand. I would expect something more natural, like:

```python
includes, vars, config = profile_split(text)
# includes is a list, vars is a dict, config is the remaining text but still text
```

**Comment by lasote** [comment]
Updated

File: conans/client/profile_loader.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+            elif line.strip().startswith("["):
+                return ret
+            elif line.strip().startswith("include("):
+                include = line.split("include(", 1)[1]
+                if not include.endswith(")"):
+                    raise ConanException("Invalid include statement")
+                include = include[:-1]
+                ret.append(include)
+            else:
+                continue
```
Comment (memsharded) [inline conans/client/profile_loader.py]: Else continue??? :P

File: conans/client/profile_loader.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+                name, value = line.split("=", 1)
+                name = name.strip()
+                if " " in name:
+                    raise ConanException("The names of the variables cannot contain spaces")
+                value = unquote(value)
+                vars[name] = value
+
+        return vars
+
+    def _parse_includes(self, text):
```
Comment (memsharded) [inline conans/client/profile_loader.py]: I still don't get why you parse 3 times the same file.

**Pull request merged**